### ISSUE TITLE
Add Advanced Search

### ISSUE THREAD

**Issue by graft** [mountetna/timur#227]
The Search page currently makes simple queries from a single table  using the Magma 'retrieve' endpoint. However, the Magma 'query' can make complex queries that can return data from anywhere across the data graph in a single table. 

We would like to add a user interface for composing complex table queries on the Search page.

By default, the user would see the simple Search, but if they click over to 'Advanced Search', they would get a new interface that allows them to build a data table by filtering and mapping columns.

The query they compose in this interface could also be copied as a 'raw query' and pasted/saved into a manifest.

**Comment by graft** [comment]
Potential layout for Advanced Search page:

    Select table_name: [ experiment ]
      (+ Add filter)
        short_name matches 'B'
        
    Columns: (+ Add column)
        abbrev_name - short_name
        first_patient_name - patient
            []
            (+ Add filter)
            sample
               []
               (+ Add filter)
               population

**Comment by dabrau** [comment]
https://balsamiq.cloud/sogp373/prswhvu

**Comment by dabrau** [comment]
screen shots of mockups. link only works with balsamiq account.
![current_search](https://user-images.githubusercontent.com/7761177/46427257-d2292880-c6f5-11e8-9924-1cc6e925a207.png)
![basic_search](https://user-images.githubusercontent.com/7761177/46427262-d6eddc80-c6f5-11e8-92f9-58965637213e.png)
![advanced_search_true_false](https://user-images.githubusercontent.com/7761177/46427268-d9e8cd00-c6f5-11e8-9e76-a3d1eba18854.png)
![advanced_search_string](https://user-images.githubusercontent.com/7761177/46427274-dead8100-c6f5-11e8-903a-fb1c2f889d34.png)
![advanced_search_string2](https://user-images.githubusercontent.com/7761177/46427278-e1a87180-c6f5-11e8-9c96-e88f1e122ce5.png)
![advanced_search_numeric](https://user-images.githubusercontent.com/7761177/46427290-e836e900-c6f5-11e8-8a77-83ff1e11cefa.png)
![advanced_search_date_time](https://user-images.githubusercontent.com/7761177/46427297-ed943380-c6f5-11e8-9a0a-524c355509d1.png)

**Comment by graft** [comment]
I like the presentation of the basic search, I think the sidebar is more consistent with the other Timur layouts. I'm not sure I understand your advanced search proposal. My essential conception of the search is there are basically three parts - the table you wish to search, filters on the rows in that table, and values mapped from each filtered row (including values from associated tables). I'm not sure I see these three parts in the advanced search interface. Could you clarify its operation?

**Comment by dabrau** [comment]
The goal I want to accomplish is giving an interface for users to pull down small sets they are interested in. In my experience pulling data for users the queries are fairly simple. Give me TPM from HEP in the tumor compartment for x, y, z genes.
- I don't think specifying a particular table makes sense if you include columns from other tables. I consider the entire set of tables as one big joined table.
- the sidebar will provide input for filtering options for the columns that are added.

**Comment by graft** [comment]
>I consider the entire set of tables as one big joined table.

From the user-query perspective the problem with this formulation (constraints are applied to the column values) is that you cannot have multiple paths through the dataset to the same column. As an example, I cannot make a table listing Lag3 MFI in CD4+ and CD8+ t-cells from the treg stain for each sample this way (or if I can I don't see how) because I wish to join MFI values to the population table in two different ways (through name='CD4+' and name='CD8+').

**Comment by dabrau** [comment]
mfi -> population -> sample
The tables are joined only by the the columns that are selected.
The assumptions that I feel confident in making is:
- The table to start the query is the farthest from the root node "Project"
- The other tables to join is the shortest path from the start table

**Comment by graft** [comment]
can you mock out how the above query would look? a table with two columns, lag3_mfi_cd4 and lag3_mfi_cd8.

**Comment by dabrau** [comment]
i don't think I would make a query like that.
I would do this:
```
@treatments = table([
  'mfi',
  ['name', '::equals', 'Lag3'],
  ['population', 'stain', '::equals', 'treg'],
  ['population', 'name', '::in', ['Q3: CD8a+,CD4-', 'Q1: CD8a-,CD4+']],
  '::all',
  [
    mfi_name: ['name'],
    mfi_value: ['value'],
    population_stain: [ 'population', 'stain'],
    population_name: [ 'population', 'name'],
    populationn_name: [ 'population', 'sample', 'sample_name']
  ]
])
```

**Comment by graft** [comment]
Hmm, so that is basically the flat version of the same table. The data is the same, the format is not. I guess now the question is can every query be flattened the same way. In this case it can I think because we're mapping the same column twice. That might not always be the case - for example if I wanted to pull population.count for population.name='CD4+' and Ki67 MFI for population.name='CD8+'. You could probably manage this pulling a superset of information using the above method, or breaking this into two separate queries, so maybe this suffices.

**Comment by dabrau** [comment]
I don't think the point is to be able to do every query possible. Thats what the manifest is for?

**Comment by graft** [comment]
I agree it need not do every query possible, but it's worth understanding what is the range of queries that this interface could express. I think with a minimal amendment this interface could go much further in the kind of queries it is able to capture.

You present an interface that allows you to select columns for mapping and impose constraints on them. The amendment I would suggest imposing is to simply clarify the difference between a mapped column and a constraint by separating these out in the interface. That is, your interface, instead of being a list of columns-and-constraints, would be a list of columns, each with their own list of constraints (where constraint is [ some_model.attribute, operation, value ]).

That interface would not be terribly more complicated than this one, and it would create an important distinction in the user's mind between 'data value I am interested in' and 'constraining values in the query'. It would also allow the MFI example operation above and would be nearly as expressive as the manifest.

**Comment by dabrau** [comment]
I don't think a column with its own constraints is easy to grok. I think creating that distinction you asked for is leaking the abstraction of the query api into the interface.

**Comment by graft** [comment]
To some extent this is all about leaking the abstraction of the query api into the interface, we're just debating how much should leak. In this case I think the basic form of the query is something like "Give my X where Y", which implies we should be making an interface-level distinction between X (values) and Y (constraints). This doesn't have to mean that there are per-column constraints, but I find the interface you presented confusing because it elides this distinction. Maybe this is something to be determined by user interviews or some such?

**Comment by dabrau** [comment]
Ok, I will move forward by finding people who use the search.
Questions to ask:
- Tell me about your experience using the search page?
- Why are you on the search page?
- What do you wish you can do on the search page? How do you envision yourself accomplishing that?
